### PR TITLE
ch4: add stubs for implicit selection of VCIs

### PR DESCRIPTION
**Pull request by sagarth**
## Pull Request Description
This PR contains patches to add the following features:
* Enable implicit hashing for VCI selection
* Add hash functions to select VCI based on (comm, rank, tag), but with basic placeholder implementation.

Update: some features features have been moved to the following PRs:

- PR: https://github.com/pmodels/mpich/pull/5481

  * Use of the standard hints: `mpi_assert_no_any_tag`, `mpi_assert_no_any_source`, and `mpi_assert_allow_overtaking` to relax the selection of VCI on sender and receiver side
  * Add non-standard user hints `vci_idx_sender`, and `vci_idx_receiver` to directly select specific VCI for a comm. This feature can be used by threads to directly distribute or isolate traffic across different VCIs.
  * Add non-standard used hint `thread_id` which can control both `vci_idx_sender` and `vci_idx_receiver` at the same time. This hint can be modified at runtime by using a CVAR. Its use can also be extended to control how we assign/offload the reduction operation to different progress threads. This hint can be useful for certain applications, e.g., OneCCL.
  * Add tests cases to evaluate functional correctness of the added comm hints

* PR: https://github.com/pmodels/mpich/pull/5476
  * Bug fixes

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by wesbland** [comment]
@sagarth Looks like there's some CH3 warnings that need to be squashed.

**Comment by sagarth** [comment]
@wesbland Let me look into them.

**Comment by sagarth** [comment]
The CH3 warnings are cleared now.

File: src/include/mpir_comm.h [hunk 1]
```diff
@@ -89,6 +89,8 @@ enum MPIR_COMM_HINT_PREDEFINED_t {
     MPIR_COMM_HINT_EAGER_THRESH,        /* ch3 */
     MPIR_COMM_HINT_EAGAIN,      /* ch4:ofi */
     /* dynamic hints starts here */
+    MPIR_COMM_HINT_VCI_IDX_SENDER,
+    MPIR_COMM_HINT_VCI_IDX_RECEIVER,
```
Comment (hzhou) [inline src/include/mpir_comm.h:93]: Move above the comment `/* dynamic hints starts here */`.  Both are still "static" hints. Dynamic hints will be tracked by a dynamic variable rather than a constant index.

File: src/mpid/ch4/src/ch4_vci.h [hunk 2]
```diff
@@ -70,6 +70,142 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_get_vci(int flag, MPIR_Comm * comm_ptr,
 }
 
 #elif MPIDI_CH4_VCI_METHOD == MPICH_VCI__IMPLICIT
+/* Remove the sign bit from an integer, if present */
+MPL_STATIC_INLINE_PREFIX int MPIDI_mask_int_sign(int i)
+{
+    return ((i) & ~(1 << (sizeof(int) * 8 - 1)));
+}
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_vci.h:77]: This is only used with `% MPIDI_global.n_vcis`, right? How about move the `% MPIDI_global.n_vcis` into this function and rename it to `MPIDI_int_to_vci`?

File: src/mpid/ch4/src/ch4_vci.h [hunk 3]
```diff
@@ -70,6 +70,142 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_get_vci(int flag, MPIR_Comm * comm_ptr,
...
+                                                  int sender_rank, int receiver_rank, int tag)
+{
+#if MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__VCI
+    MPIR_Assert(comm);
+    int vci_idx = MPIDI_VCI_INVALID;
+    bool use_user_defined_vci = (comm->hints[MPIR_COMM_HINT_VCI_IDX_SENDER] != MPIDI_VCI_INVALID);
+    bool use_tag = comm->hints[MPIR_COMM_HINT_NO_ANY_TAG];
+
+    /* Compute vci_idx if user did not provide it */
+    if (unlikely(!use_user_defined_vci)) {
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_vci.h:155]: This is not an unlikely condition. We do not optimize special cases by hurting common cases.

File: src/mpid/ch4/src/ch4_vci.h [hunk 4]
```diff
@@ -70,6 +70,142 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_get_vci(int flag, MPIR_Comm * comm_ptr,
...
+{
+#if MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__VCI
+    MPIR_Assert(comm);
+    int vci_idx = MPIDI_VCI_INVALID;
+    bool use_user_defined_vci = (comm->hints[MPIR_COMM_HINT_VCI_IDX_RECEIVER] != MPIDI_VCI_INVALID);
+    bool use_tag = comm->hints[MPIR_COMM_HINT_NO_ANY_TAG];
+    bool use_source = comm->hints[MPIR_COMM_HINT_NO_ANY_SOURCE];
+
+    /* Compute vci_idx if user did not provide it */
+    if (unlikely(!use_user_defined_vci)) {
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_vci.h:199]: Remove `unlikely`

File: src/mpid/ch4/src/ch4_vci.h [hunk 5]
```diff
@@ -70,6 +70,142 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_get_vci(int flag, MPIR_Comm * comm_ptr,
...
+    return (MPIDI_mask_int_sign(MPIR_CONTEXT_READ_FIELD(PREFIX, context_id) + tag)
+            % MPIDI_global.n_vcis);
+}
+
+/* Map comm, target_rank, and tag to vci_idx */
+MPL_STATIC_INLINE_PREFIX int MPIDI_map_contextid_rank_tag_to_vci(MPIR_Context_id_t context_id,
+                                                                 int target_rank, int tag)
+{
+    return (MPIDI_mask_int_sign(MPIR_CONTEXT_READ_FIELD(PREFIX, context_id) + target_rank + tag)
+            % MPIDI_global.n_vcis);
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_vci.h:106]: I think you can remove `target_` and just say `rank`.

File: src/mpid/ch4/src/ch4_vci.h [hunk 6]
```diff
@@ -70,6 +70,142 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_get_vci(int flag, MPIR_Comm * comm_ptr,
...
+    MPIR_Assert(comm);
+    int vci_idx = MPIDI_VCI_INVALID;
+    bool use_user_defined_vci = (comm->hints[MPIR_COMM_HINT_VCI_IDX_SENDER] != MPIDI_VCI_INVALID);
+    bool use_tag = comm->hints[MPIR_COMM_HINT_NO_ANY_TAG];
+
+    /* Compute vci_idx if user did not provide it */
+    if (unlikely(!use_user_defined_vci)) {
+        if (use_tag) {
+            vci_idx = MPIDI_map_contextid_rank_tag_to_vci(ctxid_in_effect, receiver_rank, tag);
+        } else {        /* General unoptimized case */
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_vci.h:160]: Move the comment to a separate line below.

File: src/mpid/ch4/src/ch4_vci.h [hunk 7]
```diff
@@ -70,6 +70,142 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_get_vci(int flag, MPIR_Comm * comm_ptr,
...
+            vci_idx = MPIDI_map_contextid_rank_to_vci(ctxid_in_effect, receiver_rank);
+        }
+    } else {    /* Read user provided value */
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_vci.h:163]: Move the comment to a separate line below. The comment applies to the "branch", not the `else` statement.

**Comment by wesbland** [comment]
test:mpich/ch4/ofi

File: src/include/mpir_thread.h [hunk 8]
```diff
@@ -95,6 +95,7 @@ extern MPID_Thread_mutex_t MPIR_THREAD_VCI_HANDLE_MUTEX;
 extern MPID_Thread_mutex_t MPIR_THREAD_VCI_CTX_MUTEX;
 extern MPID_Thread_mutex_t MPIR_THREAD_VCI_PMI_MUTEX;
 extern MPID_Thread_mutex_t MPIR_THREAD_VCI_BSEND_MUTEX;
+#define MPIR_THREAD_VCI_COMM_MUTEX(_comm_ptr) _comm_ptr->mutex
```
Comment (hzhou) [inline src/include/mpir_thread.h:98]: There is no benefit of this macro, suggest remove. Just use `comm_ptr->mutex` directly.

File: src/mpid/ch4/src/ch4_vci.h [hunk 9]
```diff
@@ -106,6 +106,25 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_map_contextid_rank_tag_to_vci(MPIR_Context_id
...
+    } else {
+        int vci_idx_new = (MPIDIG_COMM(comm, vci_idx_last_assigned) + 1)
+            % MPIDI_global.n_vcis;
+        MPIDIG_COMM(comm, vci_idx_last_assigned) = vci_idx_new;
+    }
+
+    MPID_THREAD_CS_EXIT(VCI, MPIR_THREAD_VCI_COMM_MUTEX(comm));
+    return MPIDIG_COMM(comm, vci_idx_last_assigned);
+}
+
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_vci.h:127]: I am not sure this is a good idea. How about split this commit to a separate PR so we can review it separately?

File: src/mpid/ch4/src/ch4_init.c [hunk 10]
```diff
@@ -102,6 +102,18 @@
...
+      category    : CH4
+      type        : string
+      default     : "thread_id"
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        This CVAR specifies the key for a comm info hint that can define thread ID for a communicator
+        or to control the mapping a communicator with a tx and rcv vci.
+
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_init.c:116]: > This feature can be useful to certain applications, e.g., OneCCL.

Could you elaborate on the use case?

File: src/mpi/comm/commutil.c [hunk 11]
```diff
@@ -587,6 +587,14 @@ int MPIR_Comm_create_subcomms(MPIR_Comm * comm)
         comm->node_comm->local_size = num_local;
         comm->node_comm->remote_size = num_local;
 
+        /* Copy relevant hints to node_comm */
+        comm->node_comm->hints[MPIR_COMM_HINT_VCI_IDX_SENDER] =
+            comm->hints[MPIR_COMM_HINT_VCI_IDX_SENDER];
+        comm->node_comm->hints[MPIR_COMM_HINT_VCI_IDX_RECEIVER] =
+            comm->hints[MPIR_COMM_HINT_VCI_IDX_RECEIVER];
+        comm->node_comm->hints[MPIR_COMM_HINT_THREAD_ID] = comm->hints[MPIR_COMM_HINT_THREAD_ID];
```
Comment (hzhou) [inline src/mpi/comm/commutil.c:595]: They are registered inside ch4, but leaked here in the MPIR layer. Is it possible to do this in the ch4-layer?

File: src/mpid/ch4/src/ch4_vci.h (hunk 2 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_vci.h:77]: Yes. That's a good idea.

File: src/include/mpir_thread.h (hunk 8 above)
Comment (sagarth) [reply src/include/mpir_thread.h:98]: Removing it may be an option. I think the original motivation was to make it consistent with `MPIR_THREAD_POBJ_COMM_MUTEX` and `MPIR_THREAD_POBJ_WIN_MUTEX`.

File: src/mpid/ch4/src/ch4_vci.h (hunk 9 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_vci.h:127]: Did you mean split just this one commit or the whole implementation to utilize the hint `allow_overtaking`? (this commit is a part of `allow_overtaking` based vci selection.)

File: src/mpi/comm/commutil.c (hunk 11 above)
Comment (sagarth) [reply src/mpi/comm/commutil.c:595]: That is right in on way.  However, with the current design, these hints are defined at mpir (in src/include/mpir_comm.h) but registered at ch4. So, it seemed keeping here may be okay. Do you still favor moving it?

File: src/mpid/ch4/src/ch4_init.c (hunk 10 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_init.c:116]: The initial motivation for this hint was to allow an application to smoothly switch from using IMPI to MPICH, with minimal changes. IMPI provides a similar feature to dynamically rename the key. An application can use different keys, e.g., "thread_id", "ep_idx", "comm_id", in different contexts. Or this value can change across different applications. When MPICH allows dynamic renaming, an application can run with either IMPI or MPICH without having to change the key string in the source code.

File: src/include/mpir_thread.h (hunk 8 above)
Comment (hzhou) [reply src/include/mpir_thread.h:98]: We have removed POBJ from ch4 layer completely.

File: src/mpid/ch4/src/ch4_vci.h (hunk 9 above)
Comment (hzhou) [reply src/mpid/ch4/src/ch4_vci.h:127]: Let's split the entire `allow_overtaking` feature.

File: src/mpid/ch4/src/ch4_init.c (hunk 10 above)
Comment (hzhou) [reply src/mpid/ch4/src/ch4_init.c:116]: I see. In stead of adding this CVAR, which is cumbersome for user to use anyway, let's make `thread_id`, `ep_idx`, `comm_id` aliases, i.e. check against all aliases in `comm_hint_test`.

File: src/mpi/comm/commutil.c (hunk 11 above)
Comment (hzhou) [reply src/mpi/comm/commutil.c:595]: Alright, let's keep it simple (as is) for now.

File: src/mpid/ch4/src/ch4_init.c (hunk 10 above)
Comment (hzhou) [reply src/mpid/ch4/src/ch4_init.c:116]: > the aliases can by changed freely by the applications.

Why would applications do that?

Comment (hzhou) [reply src/mpid/ch4/src/ch4_init.c:116]: Hints itself is a hack (until official API), adding more complexity to hints is over-engineering in my opinion. I think just offer a few aliases (only when there are real usages) is sufficient.

> So MPI will have 2 users – mpi4py and CCL, so flexible hint names may be useful here.

What's the issue with 2 users? They can use the same hint key, right?

File: src/mpid/ch4/src/ch4_vci.h (hunk 9 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_vci.h:103]: Removed all three patches for `allow_overtaking`. Will create another PR for them.

File: src/include/mpir_thread.h (hunk 8 above)
Comment (sagarth) [reply src/include/mpir_thread.h]: Let us move this discussion to the `allow_overtaking` PR https://github.com/pmodels/mpich/pull/5183.

File: src/mpid/ch4/src/ch4_vci.h (hunk 7 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_vci.h]: Done

File: src/mpid/ch4/src/ch4_vci.h (hunk 5 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_vci.h]: Done

File: src/mpid/ch4/src/ch4_vci.h (hunk 4 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_vci.h]: Done

File: src/mpid/ch4/src/ch4_vci.h (hunk 3 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_vci.h]: Done

File: src/mpid/ch4/src/ch4_vci.h (hunk 2 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_vci.h:77]: Done

File: src/include/mpir_comm.h (hunk 1 above)
Comment (sagarth) [reply src/include/mpir_comm.h]: Done

File: src/mpid/ch4/src/ch4_init.c (hunk 10 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_init.c:116]: Here are some updates after offline discussions: We can get rid of the CVAR - CCL should now be able to make make changes and adapt to MPICH. If needed, we could support multiple hard-coded aliases for the hint instead. However, at present, a single name for this hint should be sufficient.

Comment (sagarth) [reply src/mpid/ch4/src/ch4_init.c]: Removed the CVAR and changed the name "thread_id" to "vci".

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
@sagarth The custom test failed.

**Comment by sagarth** [comment]
> @sagarth The custom test failed.

Yes, I'll be looking into it.

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
The custom testing should go further than the last time but still expecting some failures. Let us see what Jenkins will report.

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/ch4/ofi

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi ----with-ch4-shmmods=posix
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi ----with-ch4-shmmods=posix
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4
env: MPIR_CVAR_DEBUG_CVARS=1
testlist: attr

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct  --disable-ofi-domain
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4
testlist: threads

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4
env: MPIR_CVAR_OFI_USE_PROVIDER=psm2

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
New failures after the recent rebase:
https://jenkins-pmrs.cels.anl.gov/view/maint/job/mpich-review-custom/638/
```
f77/spawn/connaccf 
f90/spawn/connaccf90
pt2pt/probe_unexp
```

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct --disable-ofi-domain
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit --with-ch4-shmmods=none --enable-ch4-mt=direct
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4
env: MPIR_CVAR_ODD_EVEN_CLIQUES=1

**Comment by hzhou** [comment]
@sagarth Can you split this PR and separate those simpler commits that doesn't involve implicit vci hashing?

**Comment by sagarth** [comment]
@hzhou Did the splitting and moved 3 commits to https://github.com/pmodels/mpich/pull/5476. Rebased this PR on top of the new PR.

File: src/mpid/ch4/netmod/ofi/ofi_events.c [hunk 12]
```diff
@@ -381,13 +381,14 @@ int MPIDI_OFI_get_huge_event(struct fi_cq_tagged_entry *wc, MPIR_Request * req)
     }
 
     int nic = 0;
-    int vni_src = recv_elem->remote_info.vni_src;
-    int vni_dst = recv_elem->remote_info.vni_dst;
+    /* This is receiver, so dst is local and src is remote */
+    int vni_local = recv_elem->remote_info.vni_dst;
+    int vni_remote = recv_elem->remote_info.vni_src;
```
Comment (hzhou) [inline src/mpid/ch4/netmod/ofi/ofi_events.c:386]: Is it necessary to introduce a new pair of names? All you need is a comment that this function is called on `dst`, right?

File: src/mpid/ch4/netmod/ofi/ofi_events.c [hunk 13]
```diff
@@ -411,17 +412,17 @@ int MPIDI_OFI_get_huge_event(struct fi_cq_tagged_entry *wc, MPIR_Request * req)
...
+        int ctx_idx = MPIDI_OFI_get_ctx_index(recv_elem->comm_ptr, vni_remote, nic);
         remote_key = recv_elem->remote_info.rma_keys[nic];
-        MPIDI_OFI_cntr_incr(recv_elem->comm_ptr, vni_src, nic);
+        MPIDI_OFI_cntr_incr(recv_elem->comm_ptr, vni_remote, nic);
         MPIDI_OFI_CALL_RETRY(fi_read(MPIDI_OFI_global.ctx[ctx_idx].tx,  /* endpoint     */
                                      (void *) ((char *) recv_buf + recv_elem->cur_offset),      /* local buffer */
                                      bytesToGet,        /* bytes        */
                                      NULL,      /* descriptor   */
-                                     MPIDI_OFI_comm_to_phys(recv_elem->comm_ptr, recv_elem->remote_info.origin_rank, nic, vni_src, vni_dst),    /* Destination  */
+                                     MPIDI_OFI_comm_to_phys(recv_elem->comm_ptr, recv_elem->remote_info.origin_rank, nic, vni_local, vni_remote),       /* Destination  */
```
Comment (hzhou) [inline src/mpid/ch4/netmod/ofi/ofi_events.c:422]: This is the only place that needs to be fixed, right?

File: src/mpid/ch4/src/ch4_comm.c [hunk 14]
```diff
@@ -575,4 +579,9 @@ int MPID_Intercomm_exchange_map(MPIR_Comm * local_comm, int local_leader, MPIR_C
 /* Register CH4-specific hints */
 static void register_comm_hints(MPIR_Comm * comm)
 {
+    /* Non-standard hints for VCI selection */
+    MPIR_Comm_register_hint(MPIR_COMM_HINT_VCI_IDX_SENDER, "vci_idx_sender",
+                            MPIDI_set_comm_hint_sender_vci, MPIR_COMM_HINT_TYPE_INT, 0, -1);
+    MPIR_Comm_register_hint(MPIR_COMM_HINT_VCI_IDX_RECEIVER, "vci_idx_receiver",
+                            MPIDI_set_comm_hint_receiver_vci, MPIR_COMM_HINT_TYPE_INT, 0, -1);
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_comm.c:587]: Isn't `sender_vci` and `receiver_vci` more straightforward to pronounce?

File: src/mpid/ch4/src/ch4_vci.h [hunk 15]
```diff
@@ -70,6 +70,141 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_get_vci(int flag, MPIR_Comm * comm_ptr,
...
+    bool use_tag = comm->hints[MPIR_COMM_HINT_NO_ANY_TAG];
+
+    /* Compute vci_idx if user did not provide it */
+    if (!use_user_defined_vci) {
+        if (use_tag) {
+            vci_idx = MPIDI_map_contextid_rank_tag_to_vci(ctxid_in_effect, receiver_rank, tag);
+        } else {
+            /* General unoptimized case */
+            vci_idx = MPIDI_map_contextid_rank_to_vci(ctxid_in_effect, receiver_rank);
+        }
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_vci.h:146]: Can we just `return comm_ptr->seq;` in this PR so that it can be tested? The implicit hashing should be added in the next PR.

File: src/mpid/ch4/src/ch4_vci.h [hunk 16]
```diff
@@ -208,11 +208,25 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_get_receiver_vci(MPIR_Comm * comm,
...
+        ctxid_in_effect = comm_ptr->recvcontext_id;
+    }
+
+    if (!(flag & 0x1)) {
+        /* src */
+        return MPIDI_get_sender_vci(comm_ptr, ctxid_in_effect, src_rank, dst_rank, tag);
+    } else {
+        /* dst */
+        return MPIDI_get_receiver_vci(comm_ptr, ctxid_in_effect, src_rank, dst_rank, tag);
+    }
```
Comment (hzhou) [inline src/mpid/ch4/src/ch4_vci.h:240]: This commit (`ch4: enable implicit hashing for vci `) should be squashed with previous one.

File: src/include/mpir_comm.h [hunk 17]
```diff
@@ -93,6 +93,7 @@ enum MPIR_COMM_HINT_PREDEFINED_t {
     MPIR_COMM_HINT_MULTI_NIC_PREF_NIC,  /* ch4:ofi */
     MPIR_COMM_HINT_VCI_IDX_SENDER,      /* ch4 */
     MPIR_COMM_HINT_VCI_IDX_RECEIVER,    /* ch4 */
+    MPIR_COMM_HINT_VCI, /* ch4. Generic hint, useful to select vci_idx, progress thread etc. */
```
Comment (hzhou) [inline src/include/mpir_comm.h:96]: Isn't this redundant from the two sender/receiver specific hints?

**Comment by hzhou** [comment]
@sagarth Also, please update the draft status if this PR is ready for review.

File: src/include/mpir_comm.h (hunk 17 above)
Comment (sagarth) [reply src/include/mpir_comm.h:96]: In a way, yes. We needed this higher level hint to support OneCCL.

File: src/mpid/ch4/netmod/ofi/init_addrxchg.c [hunk 18]
```diff
@@ -287,6 +287,8 @@ int MPIDI_OFI_addr_exchange_all_ctx(void)
             }
         }
     }
+    mpi_errno = MPIR_Barrier_allcomm_auto(comm, &errflag);
+    MPIR_ERR_CHECK(mpi_errno);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/init_addrxchg.c:291]: Yes. Always try rebase to main if possible.

File: src/mpid/ch4/netmod/ofi/ofi_events.c (hunk 12 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_events.c:386]: I don't think the name changes helped.

File: src/mpid/ch4/src/ch4_vci.h (hunk 15 above)
Comment (hzhou) [reply src/mpid/ch4/src/ch4_vci.h:146]: yes

File: src/include/mpir_comm.h (hunk 17 above)
Comment (hzhou) [reply src/include/mpir_comm.h:96]: I see. Do we have code to ensure the consistency between `MPIR_COMM_HINT_VCI` and the other two hints?

File: src/mpid/ch4/src/ch4_comm.c (hunk 14 above)
Comment (sagarth) [reply src/mpid/ch4/src/ch4_comm.c]: Renamed in the new PR: https://github.com/pmodels/mpich/pull/5481

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by sagarth** [comment]
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8 --enable-ch4-vci-method=implicit
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Review by hzhou — APPROVED** [review]
This looks okay.

**Comment by sagarth** [comment]
test:mpich/ch4/most
